### DISCUSSION TITLE
OnDeserialized dissappeared in 8?

### DISCUSSION THREAD

**Discussion by swegele** [MarimerLLC/csla#3893] [Questions] [answered]
Taking a stab at upgrading our .net 6 projects to .net 8.  
Only error so far is in a CSLA class that wraps a DataSet.  The following two methods hook into dataset events to translate them into CSLA IsDirty IsNew etc.  I call these when the object pops into existence:

```csharp
        protected override void UndoChangesComplete()
        {
            base.UndoChangesComplete();
            HookEventHandlers();
        }

        protected override void OnDeserialized(System.Runtime.Serialization.StreamingContext context)
        {
            base.OnDeserialized(context);
            HookEventHandlers();
        }

```

The error says "DataExtension.OnDeserialized(StreamContext) : no suitable method found to override
I looked in main branch and it seems like it's still there??  Is there a better way to do this now?

**Comment by rockfordlhotka** [comment]
I'm not sure what to do with a DataSet. Microsoft is in the process of deprecating and removing BinaryFormatter, and it will be gone in .NET 9. As a result, CSLA no longer supports BinaryFormatter, and the `OnDeserialized` method is part of that whole thing.

The DataSet (I think?) needs BinaryFormatter to be serialized. Perhaps I'm wrong - I haven't actually used a DataSet in a very long time.

Because, at least right now, the only serializer supported by CSLA is MobileFormatter, you should be able to override the methods from [IMobileObject](https://github.com/MarimerLLC/csla/blob/main/Source/Csla/Serialization/Mobile/IMobileObject.cs) (or equivalents from the base classes) to know when state and children have been set into your object instance on deserialization.

Right now there's no "final method" invoked once the full graph has been deserialized, but individual objects _in the graph_ can override things like `OnSetState` and `OnSetChildren` to know that they are being deserialized.

**Reply by swegele** [reply]
Sorry I wasn't clear enough.  I've got the serializing of the DataSet taken care of already (exactly like what you suggested in fact).

My question is, what events/methods on a CSLA business object would I want to tap into to do my `HookEvents` call.
The `OnDeserialized` was a place that I could count on the object "coming alive" and getting it's events hooked up.  What would be the corollary to that now?

**Reply by swegele** [reply]
> Right now there's no "final method" invoked once the full graph has been deserialized

Oh maybe you were saying "there is no corollary any more" ?  If so, then where would you recommend putting the HookEvents call to be sure that wherever the DataSetWrapper found itself it had properly done that step?

**Reply by rockfordlhotka** [reply]
How exactly are you serializing the DataSet itself? Are you overriding `OnGetState` and `OnSetState` and putting the serialized dataset into the serialization info object?

If so, then in your `OnSetState`, _after_ you have deserialized the dataset so it is available again, that's where I would think you could rehook the events.

**Reply by StefanOssendorf** [reply]
What about the `OnInitialized()` hook?

**Reply by swegele** [reply]
OK this is the sitch:

CSLA object that houses the wrapper exposing it's dataset property:
```csharp
    [Serializable()]
    public class DataExtension : BusinessBaseClasses.ChildBusinessBase<DataExtension>, IDataExtension
    {  ...

        public static readonly PropertyInfo<DataSetWrapper> TheDataSetProperty = RegisterProperty<DataSetWrapper>(c => c.TheDataSet);
        internal DataSet TheDataSet
        {
            get
            {
                return ReadProperty(TheDataSetProperty)?.DataSet;
            }
            private set 
            {
                var dataSetWrapper = DataSetWrapper.GetNew(value);
                LoadProperty(TheDataSetProperty, dataSetWrapper);
            }
        }

        public DataTable TheDataTable
        {
            get
            {
                return TheDataSet.Tables[0];
            }
        }

...

        private void HookEventHandlers()
        {
            if (TheDataTable != null)
            {
                TheDataTable.ColumnChanged -= OnColumnChanged;
                TheDataTable.ColumnChanged += OnColumnChanged;

                TheDataTable.RowDeleted -= OnRowDeleted;
                TheDataTable.RowDeleted += OnRowDeleted;
            }
        }

        protected override void UndoChangesComplete()
        {
            base.UndoChangesComplete();
            HookEventHandlers();
        }

        protected override void OnDeserialized(System.Runtime.Serialization.StreamingContext context)
        {
            base.OnDeserialized(context);
            HookEventHandlers();
        }

```

The wrapper of the dataset:
```csharp

    /// <summary>
    /// A class that wraps a DataSet for safe serializing / cloning by Csla
    /// </summary>
    [Serializable()]
    public class DataSetWrapper : IMobileObject 
    {
        private const string CONST_KEY_ENTITYXML = "datasetwrapper._entityXML";

        private DataSet mDataSet;

        internal static DataSetWrapper GetNew(DataSet theDataSet)
        {
            var newObj = new DataSetWrapper
            {
                mDataSet = theDataSet
            };

            return newObj;
        }

        public DataSet DataSet { get { return mDataSet; } }

        void IMobileObject.GetState(SerializationInfo info)
        {
            if (mDataSet is null)
            {
                info.AddValue(CONST_KEY_ENTITYXML, string.Empty);
            }
            else
            {
                XmlSerializer serializer = new XmlSerializer(mDataSet.GetType());
                using (StringWriter sw = new StringWriter())
                {
                    serializer.Serialize(sw, mDataSet);
                    info.AddValue(CONST_KEY_ENTITYXML, sw.ToString());
                }
            }
        }

        void IMobileObject.SetState(SerializationInfo info)
        {
            var xmlString = info.GetValue<string>(CONST_KEY_ENTITYXML);
            if (string.IsNullOrWhiteSpace(xmlString))
            {
                mDataSet = null;
            }
            else
            {
                XmlSerializer xmlSerializer = new XmlSerializer(typeof(DataSet));
                using (StringReader textReader = new StringReader(xmlString))
                {
                    mDataSet = (DataSet)xmlSerializer.Deserialize(textReader);
                }
            }
        }
        ...

```

So removing `OnDeserialized` from the BO (which makes total sense now that binary serialization is gone)...I need a similar place to call hook inside `DataExtension` when it is deserialized or initialized or created or whatever.  Just so happens that as of today and for the foreseeable future I will not be using a remote dataportal (only the local) but probably good to plan for it's possibility.

**Reply by rockfordlhotka** [reply, marked as answer]
The parent BO should be able to override `OnSetChildren`:

```c#
  protected override void OnSetChildren(...)
  {
    base.OnSetChildren(...);
    HookEventHandlers();
  }
```

If that doesn't work, we might have to consider something more like `OnDeserialized` in MobileFormatter.

**Reply by swegele** [reply]
> The parent BO should be able to override OnSetChildren:

OMG that is soooo stinkin obvious...(I shoulda had a V8!)  Thanks for that!

**Reply by ajohnstone-ks** [reply]
> The parent BO should be able to override `OnSetChildren`:

@rockfordlhotka  Sorry to reply to an older thread, but we're moving from 7 to 8 and have the same requirement.

However the child objects are `null` after calling `base.OnSetChildren(...)` so I'm not quite sure what to do.

```
protected override void OnSetChildren(SerializationInfo info, MobileFormatter formatter) {
	base.OnSetChildren(info, formatter);

	// ChildList1 & 2 below are null
	ChildList1.RemovingItem += ChildList1_RemovingItem;
	ChildList2.RemovingItem += ChildList2_RemovingItem;
}
```
